pkg/deployer/lib/extension/hook.go

```diff
@@ -0,0 +1,126 @@
+	// Beware: Aborting the reconciliation at this point might leave the deploy item in an inconsistent state and is discouraged.
+	BeforeAbortHook HookType = "BeforeAbort"
+
+	// Called if the deploy item is to be force-reconciled due to the force reconcile operation annotation.
+	// Beware: Aborting the reconciliation at this point might leave the deploy item in an inconsistent state and is discouraged.
+	BeforeForceReconcileHook HookType = "BeforeForceReconcile"
+
+	// Called if the deploy item is to be deleted.
+	// Beware: Aborting the reconciliation at this point might leave the deploy item in an inconsistent state and is discouraged.
+	BeforeDeleteHook HookType = "BeforeDelete"
+
+	// Called if a "normal" reconcile of the deploy item is about to happen.
+	// This is the case if a reconcile is required and the deploy item is not aborted, force-reconciled, or deleted.
+	// Beware: Aborting the reconciliation at this point might leave the deploy item in an inconsistent state and is discouraged.
+	BeforeReconcileHook HookType = "BeforeReconcile"
+
+	// Called before any of abortion, deletion, and (force-)reconciliation.
+	// This will always be called if the reconciliation has not been aborted before or during the ShouldReconcile check.
+	// Beware: Aborting the reconciliation at this point might leave the deploy item in an inconsistent state and is discouraged.
+	BeforeAnyReconcileHook HookType = "BeforeAnyReconcile"
+
+	// Called after a successful reconciliation.
+	// This will always be called at the end of reconciliation, unless an error has occurred or the reconciliation has been aborted
+	//   during the responsibility or ShouldReconcile check or by one of the earlier hooks.
+	EndHook HookType = "End"
+)
+
+// ReconcileExtensionHook represents a function which will be called when the hook is executed.
+type ReconcileExtensionHook func(context.Context, logr.Logger, *lsv1alpha1.DeployItem, *lsv1alpha1.Target, HookType) (*HookResult, error)
+
+// ReconcileExtensionHooks maps hook types to a list of hook functions.
+type ReconcileExtensionHooks map[HookType][]ReconcileExtensionHook
+
+// ReconcileExtensionHookSetup can be used to couple a hook function with the hooks it is meant for.
+type ReconcileExtensionHookSetup struct {
+	Hook      ReconcileExtensionHook
+	HookTypes []HookType
+}
+
+// ExecuteHooks executes all hooks of a given type in the order they are specified in.
+// The results of all executed hooks are aggregated using the AggregateHookResults function, except for
+//   DuringResponsibilityCheck and ShouldReconcile hooks, where AggregateHookResultsWithInvertedAbortPriority is used instead.
+// An error is returned if one of the hooks returns an error or if an unknown hook type is given.
+func (hooks ReconcileExtensionHooks) ExecuteHooks(ctx context.Context, log logr.Logger, di *lsv1alpha1.DeployItem, target *lsv1alpha1.Target, hype HookType) (*HookResult, error) {
```

Comment:
This one is intended, as `type` is a keyword in go and can't be used.

Reply:
Ah, yes, makes sense.
It is a matter of taste, but `hype`seems strange to me as this is an english word with a totally different meaning.
Would something like `ht` be a better fit, since we are already using abbreviations like `di`f or deploy items?

Reply:
I like the idea. Will rename it to `ht`.